### PR TITLE
S1226: Add support for deconstruction

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs [hunk 1]
```diff
@@ -38,38 +38,27 @@ public abstract class ParameterAssignedToBase<TSyntaxKind, TIdentifierNameSyntax
...
-        private bool IsReadBefore(SemanticModel semanticModel, ISymbol parameterSymbol, SyntaxNode assignment)
-        {
-            // Same problem as in VB.NET / IsAssignmentToCatchVariable:
-            // parameterSymbol.DeclaringSyntaxReferences is empty for Catch syntax in VB.NET as well as for indexer syntax for C#
-            // https://github.com/dotnet/roslyn/issues/6209
-            var stopLocation = parameterSymbol.Locations.FirstOrDefault();
-            if (stopLocation == null)
-            {
-                return true; // If we can't find the location, it's going to be FN
-            }
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs:67]: Code coverage has shown, that this code path is not hit anymore and so I removed it. The bug is fixed in newer Roslyn versions. I can't tell if it is really safe to remove the check.

Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs:67]: @pavel-mikula-sonarsource  Maybe this should be kept for backward compatibility?

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM

File: analyzers/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/ParameterAssignedToBase.cs:67]: We need to keep it as long as we support the old version, unfortunately.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterAssignedToTest.cs [hunk 2]
```diff
@@ -30,21 +30,21 @@ public class ParameterAssignedToTest
...
 
         [TestMethod]
-        public void ParameterAssignedTo_CSharp() =>
+        public void ParameterAssignedTo_CS() =>
             verifierCS.AddPaths("ParameterAssignedTo.cs").Verify();
 
 #if NET
         [TestMethod]
-        public void ParameterAssignedTo_CSharp9() =>
+        public void ParameterAssignedTo_CS9() =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterAssignedToTest.cs:38]: Not here :)
```suggestion
        public void ParameterAssignedTo_CSharp9() =>
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterAssignedToTest.cs [hunk 3]
```diff
@@ -30,21 +30,21 @@ public class ParameterAssignedToTest
...
             verifierCS.AddPaths("ParameterAssignedTo.CSharp9.cs").WithTopLevelStatements().Verify();
 
         [TestMethod]
-        public void ParameterAssignedTo_CSharp10() =>
+        public void ParameterAssignedTo_CS10() =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/ParameterAssignedToTest.cs:42]: Not here :)
```suggestion
        public void ParameterAssignedTo_CSharp10() =>
```

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM. The version-specific tests should not be renamed

**Comment by martin-strecker-sonarsource** [comment]
> LGTM. The version-specific tests should not be renamed

I fixed the test names, but merging is blocked (low coverage). Can you please merge?

**Comment by martin-strecker-sonarsource** [comment]
There are two issues marked as resolved, that are TP IMO:

https://peach.sonarsource.com/issues?open=AWc3R5Th97K2b7lhtv20&resolutions=FIXED&rules=csharpsquid%3AS1226

The history of the issues is odd.

@andrei-epure-sonarsource Can we have a look together?

**Pull request merged**